spec/message.abnf

```diff
@@ -0,0 +1,52 @@
+message = [s] *(declaration [s]) body [s]
+
+declaration = "let" s variable [s] "=" [s] "{" [s] expression [s] "}"
+body = pattern
+     / (selectors 1*([s] variant))
+
+pattern = "{" *(text / placeholder) "}"
+selectors = "match" 1*([s] selector)
+selector = "{" [s] expression [s] "}"
+variant = "when" 1*(s key) [s] pattern
+key = nmtoken / literal / "*"
+
+placeholder = "{" [s] (expression / markup / markup-end) [s] "}"
+
+expression = ((literal / variable) [s annotation])
+           / annotation
+annotation = function *(s option)
+option = name [s] "=" [s] (literal / nmtoken / variable)
+
+markup = markup-start *(s option)
+
+any-char = %x0-27 ; omit ( and )
+         / %x2A-5B ; omit \
+		 / %x5D-7A ; omit {
+		 / %x7C ; omit }
+		 / %x7E-D7FF
+		 / %xE000-10FFFF
+
+text = 1*(text-char / text-escape)
+text-char = any-char / "(" / ")"
+
+variable = "$" name
+function = ":" name
+markup-start = "+" name
+markup-end = "-" name
+name = name-start *name-char
+nmtoken = 1*name-char
```

Comment:
`name` and `nmtoken` are taken directly from https://www.w3.org/TR/xml/#NT-Name. And yes, the goal was to make it convenient to use (a) numerical option values (`fractionDigits=1`) and (b) option values such as `year=2-digit`.

More broadly, the use of `nmtoken` is intended to align MF2 with LDML. CLDR data is defined as attribute enums (e.g. in [ldmlSupplemental.dtd](https://github.com/unicode-org/cldr/blob/release-42/common/dtd/ldmlSupplemental.dtd)), which [implies that they are Nmtokens](https://www.w3.org/TR/xml/#sec-attribute-types).

Reply:
`nmtoken` however is actually exposed to more than just LDML/CLDR constructs. Developers of selectors and formatters will use them to define key values and option values. That is, our user base is wider than just CLDR stuff or ICU developers. If we define values that are a strict superset of those restrictions then any XML, LDML, or CLDR values will work. LDML being a dialect of XML means that they are restricted by XML, but we don't ***have*** to be. Thoughts?

Reply:
Wait, are you suggesting that keys and option values should be even less restricted than `nmtoken`?

Reply:
My goal here (last year when I wrote this for my proposal) was to make sure that whatever appears in LDML attlists (now or in the future) can be a valid key or an option value. That's why I chose `Nmtoken` for these two productions.

Reply:
Yes. I'm suggesting that any characters we omit we know why we are omitting them. Some implementations will probably not check the values so long as they are able to parse the various tokens (which don't really depend on the characters inside the token); others will pedantically check the character ranges.

To be honest, most developers will write enumerated values in ASCII for keys and option values. But for cases when they don't the restrictions on users should make sense in our world (and not just because someone else did it)

Reply:
@aphillips 
> LDML being a dialect of XML

I hope you meant "LDML being an application of XML". XML doesn't have any dialects (except for XML 1.1, maybe).

Reply:
@duerst exactly so